src/dodal/devices/i04/transfocator.py

```diff
@@ -18,65 +22,80 @@ class Transfocator(Device):
         my_transfocator.set(vert_beamsize_microns)
     """
 
-    beamsize_set_microns = Cpt(EpicsSignal, "VERT_REQ", kind=Kind.hinted)
-    predicted_vertical_num_lenses = Cpt(EpicsSignal, "LENS_PRED")
+    def __init__(self, prefix: str, name: str = ""):
+        with self.add_children_as_readables():
+            self.beamsize_set_microns = epics_signal_rw(float, prefix + "VERT_REQ")
+            self.predicted_vertical_num_lenses = epics_signal_rw(
+                float, prefix + "LENS_PRED"
+            )
+            self.number_filters_sp = epics_signal_rw(int, prefix + "NUM_FILTERS")
+
+            self.start = epics_signal_rw(int, prefix + "START.PROC")
+            self.start_rbv = epics_signal_r(int, prefix + "START_RBV")
 
-    number_filters_sp = Cpt(EpicsSignal, "NUM_FILTERS")
+            self.vertical_lens_rbv = epics_signal_r(float, prefix + "VER")
 
-    start = Cpt(EpicsSignal, "START.PROC")
-    start_rbv = Cpt(EpicsSignalRO, "START_RBV")
+        self.TIMEOUT = 120
+        self._POLLING_WAIT = 0.01
 
-    vertical_lens_rbv = Cpt(EpicsSignalRO, "VER", kind=Kind.hinted)
+        super().__init__(name=name)
 
-    TIMEOUT = 120
-    _POLLING_WAIT = 0.01
+    async def _observe_beamsize_microns(self):
+        have_we_done_it = False
 
-    def polling_wait_on_start_rbv(self, for_value):
+        async def set_based_on_predicition(value: float):
+            if not math.isclose(
+                self.latest_pred_vertical_num_lenses, value, abs_tol=1e-8
+            ):
+                # We can only put an integer number of lenses in the beam but the
+                # calculation in the IOC returns the theoretical float number of lenses
+                nonlocal have_we_done_it
+                value = round(value)
+                LOGGER.info(f"Transfocator setting {value} filters")
+                await self.number_filters_sp.set(value)
+                await self.start.set(1)
+                await self.polling_wait_on_start_rbv(1)
+                await self.polling_wait_on_start_rbv(0)
+                self.latest_pred_vertical_num_lenses = value
+                have_we_done_it = True
+
+        # The value hasn't changed so assume the device is already set up correctly
+        async for value in observe_value(self.predicted_vertical_num_lenses):
+            await set_based_on_predicition(value)
+            if have_we_done_it:
+                break
+
+    async def polling_wait_on_start_rbv(self, for_value):
```

Comment:
This entire function looks like it could just be replaced with `ophyd_async.core.wait_for_value`. From the comment, it looks like there were some issues with monitoring this PV in ophyd, but I wonder if it works in `ophyd-async?` 

If we do try `wait_for_value` here instead, we should go to i04 and test it before merging

Reply:
Looks like these changes worked fine on i04; resolving for now.